sycl/source/detail/graph_impl.cpp

```diff
@@ -1155,9 +1156,48 @@ void exec_graph_impl::duplicateNodes() {
   MNodeStorage.insert(MNodeStorage.begin(), NewNodes.begin(), NewNodes.end());
 }
 
+void exec_graph_impl::update(std::shared_ptr<graph_impl> GraphImpl) {
+
+  if (MDevice != GraphImpl->getDevice()) {
+    throw sycl::exception(
+        sycl::make_error_code(errc::invalid),
+        "The graphs must have been created with matching devices.");
+  }
+  if (MContext != GraphImpl->getContext()) {
+    throw sycl::exception(
+        sycl::make_error_code(errc::invalid),
+        "The graphs must have been created with matching contexts.");
+  }
+
+  if (MNodeStorage.size() != GraphImpl->MNodeStorage.size()) {
+    throw sycl::exception(sycl::make_error_code(errc::invalid),
+                          "Mismatch found in the number of nodes. The graphs "
+                          "must have a matching topology.");
+  } else {
+    for (uint32_t i = 0; i < MNodeStorage.size(); ++i) {
+      if (MNodeStorage[i]->MSuccessors.size() !=
+              GraphImpl->MNodeStorage[i]->MSuccessors.size() ||
+          MNodeStorage[i]->MPredecessors.size() !=
+              GraphImpl->MNodeStorage[i]->MPredecessors.size()) {
+        throw sycl::exception(sycl::make_error_code(errc::invalid),
+                              "Mismatch found in the number of edges. The "
+                              "graphs must have a matching topology.");
```

Comment:
Perhaps we should also be checking the type of the nodes is the same?

Reply:
Added an extra check for the type of nodes. 

Also update the line that Ewan mentioned. But didn't remove it completely because I thought it was still useful to mention the workaround.